elyra/pipeline/component_catalog.py

```diff
@@ -195,46 +352,154 @@ class ComponentCache(SingletonConfigurable):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
-        self._cache_task_queue = Queue()
-
-        # Start a thread to manage updates to the component cache
-        self._cache_manager = CacheUpdateManagerThread(self.log, self._component_cache, self._cache_task_queue)
-        self._cache_manager.name = "CacheUpdateManager"
-        self._cache_manager.start()
+        self.is_server_process = ComponentCache._determine_server_process(**kwargs)
+        self.manifest_dir = jupyter_runtime_dir()
+        # Ensure queue attribute exists for non-server instances as well.
+        self.refresh_queue: Optional[RefreshQueue] = None
+        self.update_queue: Optional[UpdateQueue] = None
+        if self.is_server_process:
+            self.refresh_queue = RefreshQueue()
+            self.update_queue = UpdateQueue()
+
+            # Set up watchdog for manifest file for out-of-process updates
+            self.observer = Observer()
+            self.observer.schedule(ManifestFileChangeHandler(self), self.manifest_dir)
+
+            # Start a thread to manage updates to the component cache
+            manager = CacheUpdateManager(
+                self.log,
+                self._component_cache,
+                self.refresh_queue,
+                self.update_queue
+            )
+            self.cache_manager = manager
+            self.cache_manager.start()
+            self.log.debug("CacheUpdateManager started...")
+        else:
+            self.manifest_filename = os.path.join(self.manifest_dir, f"elyra-component-manifest-{os.getpid()}.json")
 
-        self._build_cache()
+    @staticmethod
+    def _determine_server_process(**kwargs) -> bool:
+        """Determines if this process is a server (extension) process."""
+        app_names = ['ServerApp', 'ElyraApp']
+        is_server_process = False
+        if 'parent' in kwargs and kwargs['parent'].__class__.__name__ in app_names:
+            is_server_process = True
+        elif 'emulate_server_app' in kwargs and kwargs['emulate_server_app']:  # Used in unittests
+            is_server_process = True
+
+        return is_server_process
```

Comment:
This is proving to be a difficult thing to get working with tests, namely those using the `fp_fetch` fixture used in handler tests.  I'm finding that for the "servers" started (indirectly) by that fixture, the `ComponentCache` instance does not reflect its a "server process" when, in fact, it can ONLY be a server process.

When tests are run individually, it behaves correctly, but I'm suspicious of the fixture's teardown is not adequately cleaning things up.  I will continue looking into this, but, for now, the `test_double_refresh` test in the pipeline's `test_handlers` tests is failing.  Things work as expected in "real life".

Reply:
A single test has been found to side-affect the `test_double_refresh`.  That test is `elyra/tests/cli/test_pipeline_app.py::test_validate_with_missing_kfp_component` and there's no rhyme or reason why that particular test would do this (that I've found).  This test was reworked in #2582 so perhaps it resolves this.

For the time being, `test_double_refresh` will be marked with `xfail` so that the overall results will pass.  If the test is no longer side-affected, it will produce an `xpass` indication, at which time we can remove the marking.

Reply:
Looks like the changes in #2582 have resolved the need to mark `test_double_refresh` with `xfail` and removed via https://github.com/elyra-ai/elyra/pull/2522/commits/dbe585e976fd7624f8a64b65f9a9fa7cccabc6ca.

Reply:
After merging with latest master, the `test_double_refresh` test passes with `xpass`, so the `xfail` marker has now been removed.